### ISSUE TITLE
Rockylinux node cannot join the control plane.

### ISSUE THREAD

**Issue by git-ogawa** [git-ogawa/setup_docker#9]
## Description

In HA cluster, the rockylinux node as control plane cannot join the control plane on ubuntu node.

## Reproduction steps

Versions

- kubeadm: v1.28.2
- kubelet: v1.28.2

Use three nodes.

- ubuntu 23.04 (the first node for control plane)
- rockylinux 9.2 (the second node for control plane)
- A node for load balancer

Set the nodes as the following order in inventory_ha,yml.
```yml
  children:
    cluster:
      vars:
        ansible_ssh_port: 22
        ansible_ssh_private_key_file: ~/.ssh/id_rsa
        ansible_user: ubuntu
      children:
        control_plane:
        worker:
    control_plane:
      hosts:
        # ubuntu 23.04 node
        k8s-m1:
          ansible_host: 10.0.0.10
        # rockylinux 9.2 node
        k8s-m2:
          ansible_host: 10.0.0.11
          ansible_user: rocky
    load_balancer:
      hosts:
        load-balancer-1:
          ansible_host: 10.0.0.15
          dns_name: my-load-balancer.domain.com
```

Then run `ansible-playbook setup.yml -i inventory_ha.yml` to create k8s cluster.
The first ubuntu node is succefully created as control plane but the rockylinux node cannot join the control plane.
The bootstrap process cannot finish on the node and  the playbook fails to complete properly.

**Comment by git-ogawa** [comment]
## Investigation

On the configuration above, the ubuntu node become the first node of control plane by `kubeadm init` at first and then the rockylinux node will join the control plane by `kubeadm join` command.
On the ubuntu node, the resolvConf of kubelet is set to `/run/systemd/resolve/resolv.conf` in after `kubeadm init` succefully completed.

```
ubuntu@k8s-m1:~$ sudo cat /var/lib/kubelet/config.yaml | grep resolvConf
resolvConf: /run/systemd/resolve/resolv.conf
```

The file is managed by `systemd-reoslved` service.
```
ubuntu@k8s-m1:~$ ls -l /run/systemd/resolve/resolv.conf
-rw-r--r-- 1 systemd-resolve systemd-resolve 795 Jan 19 14:53 /run/systemd/resolve/resolv.conf

ubuntu@k8s-m1:~$ sudo systemctl status systemd-resolved.service
● systemd-resolved.service - Network Name Resolution
     Loaded: loaded (/lib/systemd/system/systemd-resolved.service; enabled; preset: enabled)
     Active: active (running) since Fri 2024-01-19 14:53:23 UTC; 19h ago
       Docs: man:systemd-resolved.service(8)
             man:org.freedesktop.resolve1(5)
             https://www.freedesktop.org/wiki/Software/systemd/writing-network-configuration-managers
             https://www.freedesktop.org/wiki/Software/systemd/writing-resolver-clients
   Main PID: 453074 (systemd-resolve)
     Status: "Processing requests..."
      Tasks: 1 (limit: 4642)
     Memory: 2.7M
        CPU: 313ms
     CGroup: /system.slice/systemd-resolved.service
             └─453074 /lib/systemd/systemd-resolved
```

When a node join to control plane by `kubeadm joint`, the kubelet configuration is derived from the first node (ubuntu node here).
However the file not found on the rockylinux node. `systemd-reoslved` service is not installed by default and only `/etc/resolv.conf` exists on the node.
```
[rocky@k8s-m2 ~]$ ls -l /run/systemd/resolve/resolv.conf
ls: cannot access '/run/systemd/resolve/resolv.conf': No such file or directory

[rocky@k8s-m2 ~]$ sudo systemctl status systemd-resolved.service
Unit systemd-resolved.service could not be found.

[rocky@k8s-m2 ~]$ ls -l /etc/resolv.conf
-rw-r--r--. 1 root root 90 Jan 18 12:57 /etc/resolv.conf
```

So the pods required for control plane cannot be created due to the reason above.
The error about the details is output in `kubelet` log on the node.
```
dns.go:284] "Could not open resolv conf file." err="open /run/systemd/resolve/resolv.conf: no such file or directory"
Jan 18 21:30:08 k8s-m2 kubelet[2181]: E0118 21:30:08.148696    2181 kuberuntime_sandbox.go:45] "Failed to generate sandbox config for pod" err="open /run/systemd/resolve/resolv.conf: no suc
```

In summary, the difference of resolv.conf path between on ubuntu and rockylinux causes the issue.

**Comment by git-ogawa** [comment]
To change resolv.conf path from the default configuration, pass `--resolv-conf` args to kubelet as mentioned in [kubernetes documantation](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/kubelet-integration/#providing-instance-specific-configuration-details).
In kubeadm, this is applied by settings `NodeRegistrationOptions.kubeletExtraArgs` in [kubeadm configuration](https://kubernetes.io/docs/reference/config-api/kubeadm-config.v1beta3/) as below.

```yml
apiVersion: kubeadm.k8s.io/v1beta3
kind: JoinConfiguration
nodeRegistration:
  kubeletExtraArgs:
    resolv-conf: /etc/resolv.conf
```

Then add a node to the cluster with `kubeadm join --config [config_path]`.

To take above method, the rockylinux node successfully join the control plane and `--resolv-conf` is set in kubelet process. 
```
[rocky@k8s-m2 ~]$ ps aux | grep kubelet
root       33749  1.5  2.9 1635636 112140 ?      Ssl  14:42   0:05 /usr/bin/kubelet --bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf --config=/var/lib/kubelet/config.yaml --container-runtime-endpoint=unix:///var/run/containerd/containerd.sock --pod-infra-container-image=registry.k8s.io/pause:3.9 --resolv-conf=/etc/resolv.conf
```

`--resolv-conf` setting is added in `/var/lib/kubelet/kubeadm-flags.env`.
```
[rocky@k8s-m2 ~]$ sudo cat /var/lib/kubelet/kubeadm-flags.env
KUBELET_KUBEADM_ARGS="--container-runtime-endpoint=unix:///var/run/containerd/containerd.sock --pod-infra-container-image=registry.k8s.io/pause:3.9 --resolv-conf=/etc/resolv.conf"
```

**Comment by git-ogawa** [comment]
Fix on https://github.com/git-ogawa/setup_kube_cluster/pull/11/commits/f0998bd30fde571d2f409aac531bb1955da0c922 and merge on https://github.com/git-ogawa/setup_kube_cluster/pull/11
